In [3]:
# Import libraries and load data from PostgreSQL
import pandas as pd
import numpy as np
from sqlalchemy import create_engine

# Connect to PostgreSQL
engine = create_engine(
    'postgresql://postgres:1234@localhost:5432/corp_card_fraud'
)

# Load processed data from PostgreSQL
df = pd.read_sql('SELECT * FROM processed_transactions', engine)         #pd.read_sql = reads data FROM PostgreSQL into pandas dataframe

print("Data loaded from PostgreSQL!")
print("Shape:", df.shape)
print("Columns:", df.columns.tolist())
print("Class distribution:\n", df['Class'].value_counts())

Data loaded from PostgreSQL!
Shape: (284807, 31)
Columns: ['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Class', 'Hour', 'Amount_Scaled']
Class distribution:
 Class
0    284315
1       492
Name: count, dtype: int64


In [4]:
# X = all columns except Class (input features)
X = df.drop('Class', axis=1)

# y = only Class column (target)
y = df['Class'] 


print("X shape:", X.shape)
print("y shape:", y.shape)
print("Features:", X.columns.tolist())

X shape: (284807, 30)
y shape: (284807,)
Features: ['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Hour', 'Amount_Scaled']


In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,      # 20% for testing , 80% for training
    random_state=42,    # reproducible results
    stratify=y          # # Maintains same fraud ratio in both train and test sets
)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)
print("\nTraining class distribution:\n", y_train.value_counts())
print("\nTest class distribution:\n", y_test.value_counts())

X_train shape: (227845, 30)
X_test shape: (56962, 30)
y_train shape: (227845,)
y_test shape: (56962,)

Training class distribution:
 Class
0    227451
1       394
Name: count, dtype: int64

Test class distribution:
 Class
0    56864
1       98
Name: count, dtype: int64


**MY REFERENCE: How sampling_strategy works:**
```
sampling_strategy = fraud / legitimate

50/50:
227,451 / 227,451 = 1.0
smote = SMOTE(sampling_strategy=1.0)

40/60:
151,634 / 227,451 = 0.67
smote = SMOTE(sampling_strategy=0.67)

30/70:
97,479 / 227,451 = 0.43
smote = SMOTE(sampling_strategy=0.43)

In [6]:
from imblearn.over_sampling import SMOTE

# Apply SMOTE to training data ONLY

#smote = SMOTE(random_state=42)                             #default case will 50-50 when we use SMOTE
smote = SMOTE(sampling_strategy=1.0, random_state=42)       # Keeping it 50-50 to maintain balance, so that model will not learn only legitimate or only fraud 
X_train_smote, y_train_smote = smote.fit_resample(X_train, y_train)

print("Before SMOTE:")
print("Legitimate:", sum(y_train==0))
print("Fraud:", sum(y_train==1))

print("\nAfter SMOTE:")
print("Legitimate:", sum(y_train_smote==0))
print("Fraud:", sum(y_train_smote==1))

Before SMOTE:
Legitimate: 227451
Fraud: 394

After SMOTE:
Legitimate: 227451
Fraud: 227451


FOR MY REFERENCE:
===================
how SMOTE works?
=================

SMOTE works using a technique called KNN (K-Nearest Neighbors)
SMOTE:
1. Take real fraud transaction
2. Find K nearest fraud neighbors
3. Pick random neighbor
4. Create new point between them
5. Repeat until balanced 50/50


EXAMPLE
=========
Real fraud transactions:

T1: V1=1.2, V17=-5.2, Amount=122

T2: V1=1.5, V17=-4.8, Amount=145

T3: V1=0.9, V17=-6.1, Amount=98

T4: V1=1.8, V17=-4.2, Amount=167

T5: V1=1.1, V17=-5.5, Amount=110

//-------------------------------------------------------------------------------

For T1 (V1=1.2, V17=-5.2, Amount=122) - Which looks most similar to T1

Find 5 closest fraud transactions:

Closest = T5 (V1=1.1, V17=-5.5, Amount=110)

Second closest  = T3 (V1=0.9, V17=-6.1, Amount=98)

Pick random neighbor - T5:

T1:  V1=1.2, V17=-5.2, Amount=122

T5:  V1=1.1, V17=-5.5, Amount=110

//-------------------------------------------------------------------------------

Pick random point: random = 0.7 (between 0 and 1)

//-------------------------------------------------------------------------------

New fake transaction: 

Formula= (t1.v1 + random) * (t5.v1 -t1.v1) 

V1     = 1.2 + 0.7 × (1.1 - 1.2) = 1.13      

V17    = -5.2 + 0.7 × (-5.5 - -5.2) = -5.41

Amount = 122 + 0.7 × (110 - 122) = 113.6

//-------------------------------------------------------------------------------

New fake fraud transaction: 

V1=1.13, V17=-5.41, Amount=113.6




In [7]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score 
# classification_report = shows Precision, Recall, F1, for each class in one table                        

# confusion_matrix = shows:
#  True Positives, False Positives
#  True Negatives, False Negatives

# roc_auc_score = calculates AUC-ROC score
#                 overall model quality (0.5 to 1.0)
import time

print("Training Logistic Regression...")
start = time.time()

# Train model
lr_model = LogisticRegression(
    class_weight='balanced',  # extra safety for imbalance 
    random_state=42,
    max_iter=1000             
    # every iterations, weight adjustments will happen! 
    # #Wrong? → adjust weights → try again 
    # Right? → keep weights   → move on
)

lr_model.fit(X_train_smote, y_train_smote)

end = time.time()
print(f"Training complete in {round(end-start, 2)} seconds!")

Training Logistic Regression...
Training complete in 10.73 seconds!


For my reference on  class_weight='balanced'
=============================================

Class weight = total samples / (number of classes × samples in that class)

Total samples     = 284,807

Number of classes = 2 (fraud and legitimate)

//-------------------------

Legitimate weight:

= 284,807 / (2 × 284,315)

= 284,807 / 568,630

= 0.50  ← low weight (majority)

//-------------------------

Fraud weight:

= 284,807 / (2 × 492)

= 284,807 / 984

= 289.4  ← high weight (minority)

So fraud gets 289x more attention than legitimate — automatically! 

//--------------------------------------------

# Manual weights — focus on legitimate

class_weight = {0: 289.4, 1: 0.50}

# class_weight='balanced' -recommendable! why? it concentrates on what is needed

How does it know which one to concentrate on? 

Legitimate = 284,807 / (2 × 284,315) = 0.50

Fraud      = 284,807 / (2 × 492)     = 289.4

Higher weight → more concentration

Lower weight  → less concentration

Fraud gets 289x more attention

because it's minority!



In [8]:
# Make predictions on TEST data
y_pred_lr = lr_model.predict(X_test)
y_pred_lr_prob = lr_model.predict_proba(X_test)[:, 1]

# Print evaluation metrics
print("LOGISTIC REGRESSION RESULTS")
print("=" * 50)
print("\nClassification Report:")
print(classification_report(y_test, y_pred_lr))
print(f"AUC-ROC Score: {roc_auc_score(y_test, y_pred_lr_prob):.4f}")

LOGISTIC REGRESSION RESULTS

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.97      0.99     56864
           1       0.05      0.92      0.10        98

    accuracy                           0.97     56962
   macro avg       0.53      0.95      0.54     56962
weighted avg       1.00      0.97      0.98     56962

AUC-ROC Score: 0.9715


# My understanding

AUC-ROC Score: 0.9715 out of 1.0
Class 0 (Legitimate):

Precision = 1.00 - perfect! 

Recall    = 0.97 0 - catches 97% of legitimate

F1        = 0.99 

Class 1 (Fraud):

Precision = 0.05 - very low! 

Recall    = 0.92 - catches 92% of fraud 

F1        = 0.10 - low

//==============================================================================

Support:

Just shows how many of each class in test data:

Class 0: 56,864 legitimate transactions

Class 1: 98 fraud transactions

//==============================================================================

 Precision :

Class 0 Precision = 1.00:

Model identified "legitimate" were actually legitimate 


Class 1 Precision = 0.05:

Model identified "fraud" but Only 5% correct! 

//==============================================================================
 
Recall- Of all ACTUAL fraud/legitimate cases, how many did model catch?

Simple understanding:

There are 100 transactions, model caught 92 of them.

Recall = 92/100 = 92% 

Formula:

# Recall = True Positives / (True Positives + False Negatives)

Class 0 Recall = 0.97

"Model correctly identified 97% of all legitimate transactions "

Class 1 Recall = 0.92

"Model caught 92% of all fraud cases! Only missed 8% "

//--------------------------------------------------------------------------------

F1 Score

Formula:
F1 = 2 × (Precision × Recall) / (Precision + Recall)

Class 0 F1 = 0.99

Class 1 F1 = 0.10

//------------------------------------------------------------------------------

Accuracy: "Overall how many predictions were correct?"

Formula:
Accuracy = Correct predictions /Total predictions
         = 97%

//-----------------------------------------------------------------------------

AUC-ROC: 0.9715
If I show model one fraud and one legitimate, what's the chance it ranks fraud higher?
sklearn calculates automatically, complex calculation

/-----------------------------------------------------------------------------------

Macro avg: Simple average of both classes

Precision: (1.00 + 0.05) / 2 = 0.53

Recall:    (0.97 + 0.92) / 2 = 0.95

F1:        (0.99 + 0.10) / 2 = 0.54

//-------------------------------------------------------


In [9]:
# from sklearn.svm import SVC
# import time

# print("Training SVM...")
# print("This will take 20-40 minutes...")
# start = time.time()

# svm_model = SVC(
#     kernel='rbf',
#     class_weight='balanced',
#     random_state=42,
#     probability=True
# )

# svm_model.fit(X_train_smote, y_train_smote)

# end = time.time()
# print(f"Training complete in {round(end-start, 2)} seconds!")

I initially used SVM with RBF kernel but switched to LinearSVC due to computational constraints(It took more than 3 hours n still it was running) on 454,902 rows after SMOTE — LinearSVC is more practical for large datasets while maintaining strong performance"

# My Reference: 

# probability=True

By default SVM only says , "Its fraud" or "Its legitimate" --> Just 0 or 1 — no probability!

probability=True makes SVM say

"This is 94% likely fraud" OR "This is 23% likely fraud"

Why needed? --> For AUC-ROC calculation we need probability scores not 0 or 1

# RBF

RBF = lets SVM draw any shape boundary, not just straight lines

//------------------------------------------------------------------------------------------------------

After SMOTE training data = 454,902 rows

SVM compares every transaction with every other transaction, so its slow

In [10]:
from sklearn.svm import LinearSVC
from sklearn.calibration import CalibratedClassifierCV
import time

print("Training LinearSVC...")
start = time.time()

svm_model = CalibratedClassifierCV(
    LinearSVC(
        class_weight='balanced',
        random_state=42,
        max_iter=1000
    )
)

svm_model.fit(X_train_smote, y_train_smote)

end = time.time()
print(f"Training complete in {round(end-start, 2)} seconds!")

Training LinearSVC...
Training complete in 17.46 seconds!


In [11]:
y_pred_svm = svm_model.predict(X_test)
y_pred_svm_prob = svm_model.predict_proba(X_test)[:, 1]

print("SVM RESULTS")
print("=" * 50)
print("\nClassification Report:")
print(classification_report(y_test, y_pred_svm))
print(f"AUC-ROC Score: {roc_auc_score(y_test, y_pred_svm_prob):.4f}")

SVM RESULTS

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.97      0.99     56864
           1       0.05      0.92      0.10        98

    accuracy                           0.97     56962
   macro avg       0.53      0.95      0.54     56962
weighted avg       1.00      0.97      0.98     56962

AUC-ROC Score: 0.9742


In [12]:
#training Random Forest

from sklearn.ensemble import RandomForestClassifier
import time

print("Training Random Forest...")
start = time.time()

rf_model = RandomForestClassifier(
    n_estimators=100,      # 100 decision trees
    class_weight='balanced', # focus on minority class
    random_state=42,
    n_jobs=-1              # use all CPU cores = faster!
)

rf_model.fit(X_train_smote, y_train_smote)

end = time.time()
print(f"Training complete in {round(end-start, 2)} seconds!")


Training Random Forest...
Training complete in 83.06 seconds!


In [13]:
#to print random_forest_results

y_pred_rf = rf_model.predict(X_test)
y_pred_rf_prob = rf_model.predict_proba(X_test)[:, 1]

print("RANDOM FOREST RESULTS")
print("=" * 50)
print("\nClassification Report:")
print(classification_report(y_test, y_pred_rf))
print(f"AUC-ROC Score: {roc_auc_score(y_test, y_pred_rf_prob):.4f}")

RANDOM FOREST RESULTS

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.86      0.83      0.84        98

    accuracy                           1.00     56962
   macro avg       0.93      0.91      0.92     56962
weighted avg       1.00      1.00      1.00     56962

AUC-ROC Score: 0.9845


In [14]:
from xgboost import XGBClassifier
import time

print("Training XGBoost...")
start = time.time()

xgb_model = XGBClassifier(
    n_estimators=100,       # 100 trees
    scale_pos_weight=289,   # handles class imbalance
                            # legitimate/fraud ratio ==> 284315/492 = 289
    random_state=42,
    eval_metric='logloss',  # suppress warnings
    verbosity=0             # no extra output
)

xgb_model.fit(X_train_smote, y_train_smote)

end = time.time()
print(f"Training complete in {round(end-start, 2)} seconds!")

Training XGBoost...
Training complete in 5.41 seconds!


In [15]:
from sklearn.metrics import classification_report, roc_auc_score

y_pred_xgb = xgb_model.predict(X_test)
y_pred_xgb_prob = xgb_model.predict_proba(X_test)[:, 1]

print("XGBOOST RESULTS")
print("=" * 50)
print("\nClassification Report:")
print(classification_report(y_test, y_pred_xgb))
print(f"AUC-ROC Score: {roc_auc_score(y_test, y_pred_xgb_prob):.4f}")

XGBOOST RESULTS

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.54      0.87      0.66        98

    accuracy                           1.00     56962
   macro avg       0.77      0.93      0.83     56962
weighted avg       1.00      1.00      1.00     56962

AUC-ROC Score: 0.9739


MY UNDERSTANDING:

Precision = 0.54 - better than LR/SVC but lower than RF 

Recall = 0.87    - catches 87% of fraud slightly better than RF (0.83)

F1 = 0.66        - better than LR/SVC but lower than RF 

AUC-ROC = 0.9739 - good but lower than RF 

Speed = 5 sec    - fastest compared to all 


In [16]:
from lightgbm import LGBMClassifier
import time

print("Training LightGBM...")
start = time.time()

lgbm_model = LGBMClassifier(
    n_estimators=100,
    scale_pos_weight=289,
    random_state=42,
    verbosity=-1          # suppress warnings
)

lgbm_model.fit(X_train_smote, y_train_smote)

end = time.time()
print(f"Training complete in {round(end-start, 2)} seconds!")

Training LightGBM...
Training complete in 8.15 seconds!


In [17]:
y_pred_lgbm = lgbm_model.predict(X_test)
y_pred_lgbm_prob = lgbm_model.predict_proba(X_test)[:, 1]

print("LIGHTGBM RESULTS")
print("=" * 50)
print("\nClassification Report:")
print(classification_report(y_test, y_pred_lgbm))
print(f"AUC-ROC Score: {roc_auc_score(y_test, y_pred_lgbm_prob):.4f}")

LIGHTGBM RESULTS

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      0.99     56864
           1       0.10      0.91      0.19        98

    accuracy                           0.99     56962
   macro avg       0.55      0.95      0.59     56962
weighted avg       1.00      0.99      0.99     56962

AUC-ROC Score: 0.9802


In [18]:

results = pd.DataFrame({
    'Model': ['Logistic Regression', 'LinearSVC', 
              'Random Forest', 'XGBoost', 'LightGBM'],
    'Precision': [0.05, 0.05, 0.86, 0.54, 0.10],
    'Recall':    [0.92, 0.92, 0.83, 0.87, 0.91],
    'F1':        [0.10, 0.10, 0.84, 0.66, 0.19],
    'AUC_ROC':   [0.9715, 0.9742, 0.9845, 0.9739, 0.9802],
    'Time':      ['12s', '20s', '103s', '5s', '8s']
})

print("FINAL MODEL COMPARISON")
print("=" * 70)
print(results.to_string(index=False))
print("\nBest Model: Random Forest")
print("- Highest Precision: 0.86")
print("- Highest F1: 0.84")
print("- Highest AUC-ROC: 0.9845")

FINAL MODEL COMPARISON
              Model  Precision  Recall   F1  AUC_ROC Time
Logistic Regression       0.05    0.92 0.10   0.9715  12s
          LinearSVC       0.05    0.92 0.10   0.9742  20s
      Random Forest       0.86    0.83 0.84   0.9845 103s
            XGBoost       0.54    0.87 0.66   0.9739   5s
           LightGBM       0.10    0.91 0.19   0.9802   8s

Best Model: Random Forest
- Highest Precision: 0.86
- Highest F1: 0.84
- Highest AUC-ROC: 0.9845


MY UNDERSTANDING

Best Precision: 0.86 

Best F1:        0.84 

Best AUC-ROC:   0.9845 

Reasonable time: 103 sec 

In [19]:
import joblib
import os

# Create models folder if not exists
os.makedirs('../models', exist_ok=True)

# Save all three models
joblib.dump(lr_model, '../models/logistic_regression.pkl')
joblib.dump(svm_model, '../models/linear_svc.pkl')
joblib.dump(rf_model, '../models/random_forest.pkl')
joblib.dump(xgb_model, '../models/xgboost.pkl')
joblib.dump(lgbm_model, '../models/lightgbm.pkl')

print("All models saved successfully!")
print("Saved:")
print("- models/logistic_regression.pkl")
print("- models/linear_svc.pkl")
print("- models/random_forest.pkl")
print("- models/xgboost.pkl")
print("- models/lightgbm.pkl")

All models saved successfully!
Saved:
- models/logistic_regression.pkl
- models/linear_svc.pkl
- models/random_forest.pkl
- models/xgboost.pkl
- models/lightgbm.pkl
